sdks/java/extensions/smb/src/main/java/org/apache/beam/sdk/extensions/smb/avro/AvroSortedBucketIO.java

```diff
@@ -0,0 +1,183 @@
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.smb.avro;
+
+import java.io.Serializable;
+import java.util.Optional;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.beam.sdk.coders.AvroCoder;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.IterableCoder;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.coders.NullableCoder;
+import org.apache.beam.sdk.extensions.smb.SMBCoGbkResult;
+import org.apache.beam.sdk.extensions.smb.SMBCoGbkResult.ToResult;
+import org.apache.beam.sdk.extensions.smb.SMBFilenamePolicy;
+import org.apache.beam.sdk.extensions.smb.SortedBucketFile;
+import org.apache.beam.sdk.extensions.smb.SortedBucketFile.Writer;
+import org.apache.beam.sdk.extensions.smb.SortedBucketSink;
+import org.apache.beam.sdk.extensions.smb.SortedBucketSource;
+import org.apache.beam.sdk.extensions.smb.SortedBucketSource.KeyedBucketSources.KeyedBucketSource;
+import org.apache.beam.sdk.io.fs.ResourceId;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.base.Supplier;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.collect.ImmutableList;
+
+/**
+ * Abstracts SMB sources and sinks for Avro-typed values.
+ *
+ * <p>Todo - use AutoValue builders
+ */
+public class AvroSortedBucketIO {
+
+  public static <SortingKeyT> SortedBucketSink<SortingKeyT, GenericRecord> sink(
+      AvroBucketMetadata<SortingKeyT, GenericRecord> bucketingMetadata,
+      ResourceId outputDirectory,
+      ResourceId tempDirectory,
+      Schema schema) {
+    return sink(bucketingMetadata, outputDirectory, tempDirectory, null, schema);
+  }
+
+  public static <SortingKeyT, ValueT extends GenericRecord>
+      SortedBucketSink<SortingKeyT, ValueT> sink(
+          AvroBucketMetadata<SortingKeyT, ValueT> bucketingMetadata,
+          ResourceId outputDirectory,
+          ResourceId tempDirectory,
+          Class<ValueT> recordClass,
+          Schema schema) {
+    return new SortedBucketSink<>(
+        bucketingMetadata,
+        new SMBFilenamePolicy(outputDirectory, "avro"),
+        new AvroWriterSupplier<>(recordClass, schema),
+        tempDirectory);
+  }
+
+  static class AvroWriterSupplier<ValueT> implements Supplier<Writer<ValueT>>, Serializable {
+    SortedBucketFile<ValueT> sortedBucketFile;
+
+    AvroWriterSupplier(Class<ValueT> recordClass, Schema schema) {
+      this.sortedBucketFile = new AvroSortedBucketFile<>(recordClass, schema);
+    }
+
+    @Override
+    public Writer<ValueT> get() {
+      return sortedBucketFile.createWriter();
+    }
+  }
+
+  /**
+   * Implements a typed SortedBucketSource for 2 sources.
+   *
+   * @param <KeyT>
+   * @param <V1>
+   * @param <V2>
+   */
+  public static class SortedBucketSourceJoinBuilder<KeyT, V1, V2> implements Serializable {
```

Comment:
This should get pulled out into the main smb package & generalized, since joining heterogenous source types is a valid use case